app/src/main/java/com/hedvig/app/feature/home/ui/changeaddress/ChangeAddressViewModel.kt

```diff
@@ -0,0 +1,75 @@
+package com.hedvig.app.feature.home.ui.changeaddress
+
+import androidx.lifecycle.LiveData
+import androidx.lifecycle.MutableLiveData
+import androidx.lifecycle.ViewModel
+import androidx.lifecycle.viewModelScope
+import com.hedvig.app.feature.home.ui.changeaddress.GetSelfChangeEligibilityUseCase.SelfChangeEligibilityResult
+import com.hedvig.app.feature.home.ui.changeaddress.GetUpcomingAgreementUseCase.UpcomingAgreementResult
+import com.hedvig.app.feature.home.ui.changeaddress.ViewState.ChangeAddressInProgress
+import com.hedvig.app.feature.home.ui.changeaddress.ViewState.Loading
+import com.hedvig.app.feature.home.ui.changeaddress.ViewState.ManualChangeAddress
+import com.hedvig.app.feature.home.ui.changeaddress.ViewState.SelfChangeAddress
+import com.hedvig.app.feature.home.ui.changeaddress.ViewState.SelfChangeError
+import com.hedvig.app.feature.home.ui.changeaddress.ViewState.UpcomingAgreementError
+import kotlinx.coroutines.launch
+
+abstract class ChangeAddressViewModel: ViewModel() {
+    abstract val viewState: LiveData<ViewState>
+    abstract fun reload()
+}
+
+class ChangeAddressViewModelImpl(
+    private val getUpcomingAgreement: GetUpcomingAgreementUseCase,
+    private val getSelfChangeEligibility: GetSelfChangeEligibilityUseCase,
+) : ChangeAddressViewModel() {
+
+    override val viewState: LiveData<ViewState>
+        get() = _viewState
+
+    private val _viewState = MutableLiveData<ViewState>()
+
+    init {
+        fetchDataAndCreateState()
+    }
+
+    private fun fetchDataAndCreateState() {
+        viewModelScope.launch {
+            _viewState.value = Loading
+            _viewState.value = createViewState()
+        }
+    }
+
+    private suspend fun createViewState(): ViewState {
+        return getUpComingAgreementState(
+            onNoUpcomingChange = ::getSelfChangeState
+        )
+    }
```

Comment:
```suggestion
    private suspend fun createViewState() =
        getUpComingAgreementState(
            onNoUpcomingChange = ::getSelfChangeState
        )
```

I'm partial to the functional form with an inferred return type, let me know if you feel differently 🙂

Reply:
Also, is the underlying function ever called from anywhere but here? If not, maybe we can just inline it? 😬

Reply:
Think this reads just a bit clearer tbh :)

Reply:
Fair enough!